ddtrace/profiling/collector/_traceback.pyx

```diff
@@ -49,11 +102,25 @@ cpdef pyframe_to_frames(frame, max_nframes):
     :return: The serialized frames and the number of frames present in the original traceback."""
     frames = []
     nframes = 0
-    while frame is not None:
-        nframes += 1
-        if len(frames) < max_nframes:
-            code = frame.f_code
-            lineno = 0 if frame.f_lineno is None else frame.f_lineno
-            frames.append((code.co_filename, lineno, code.co_name, _extract_class_name(frame)))
-        frame = frame.f_back
+    # Python 3.11 moved PyFrameObject members to internal C API and cannot be directly accessed
+    IF PY_MAJOR_VERSION > 3 or (PY_MAJOR_VERSION == 3 and PY_MINOR_VERSION >= 11):
+        cdef PyFrameObject* pyframe = <PyFrameObject*> frame
+        while pyframe != NULL:
+            nframes += 1
+            if len(frames) < max_nframes:
+                code = PyFrame_GetCode(pyframe)
+                lineno = PyFrame_GetLineNumber(pyframe)
```

Comment:
do you get a `PyObject*` here? does Cython handle `lineno is None` ok here, it didn't like it for `PyFrame_GetBack` though, right?

Reply:
`PyFrame_GetLineNumber` returns an int, so I think we're okay using a `is None` check. `PyFrame_GetBack` returns a pointer `PyFrameObject*` so that's why we couldn't do a `is None` check I believe.

Reply:
Note that `int` is a C `int` here. So it can't be a reference to `None`. My guess is that it would return either 0 or a negative number to signal an error

Reply:
I think we can adopt the convention that `lineno <= 0` means we failed to retrieve the line number, so we don't need to do further checks on the return value.